libnetwork/ipamutils/utils.go

```diff
@@ -71,52 +79,146 @@ func GetLocalScopeDefaultNetworks() []*net.IPNet {
+// into /96). See https://github.com/moby/moby/issues/40275.
+type NetworkSplitter struct {
+	networks []*NetworkToSplit
+	// index tracks what NetworkToSplit should be used for next allocation.
+	index int
+	// offset tracks how much of the current NetworkToSplit (indicated by index) has been consumed so far. It's
+	// reset every time index changes.
+	offset uint64
+}
 
-	for _, p := range list {
-		_, b, err := net.ParseCIDR(p.Base)
+// NewNetworkSplitter creates a NetworkSplitter from a list of NetworkToSplit. It returns an error if one of the
+// provided network is invalid (ie could not be parsed or wanted subnet Size is smaller than Base mask).
+func NewNetworkSplitter(networks []*NetworkToSplit) (*NetworkSplitter, error) {
+	for _, network := range networks {
+		_, b, err := net.ParseCIDR(network.Base)
 		if err != nil {
-			return nil, fmt.Errorf("invalid base pool %q: %v", p.Base, err)
+			return nil, fmt.Errorf("invalid base pool %q: %v", network.Base, err)
+		}
+
+		ones, bits := b.Mask.Size()
+		if network.Size <= 0 || network.Size < ones {
+			return nil, fmt.Errorf("invalid pools size: %d", network.Size)
+		}
+
+		network.baseIPNet = b
+		network.mask = net.CIDRMask(network.Size, bits)
+
+		// When there's more than 2^64 subnets that could be generated from a network, just cap maxOffset to 2^64 as
+		// this is already a number that couldn't be realistically reached.
+		if network.Size-ones > 64 {
+			network.maxOffset = 1<<64 - 1
+		} else {
+			network.maxOffset = 1 << uint(network.Size-ones)
+		}
+
+		if bits == 32 {
+			network.ipVersion = IPv4
+		} else {
+			network.ipVersion = IPv6
 		}
-		ones, _ := b.Mask.Size()
-		if p.Size <= 0 || p.Size < ones {
-			return nil, fmt.Errorf("invalid pools size: %d", p.Size)
+	}
+
+	return &NetworkSplitter{
+		networks: networks,
+	}, nil
+}
+
+// Copy returns a new NetworkSplitter with the same networks to split but with its internal position reset.
+func (ns *NetworkSplitter) Copy() *NetworkSplitter {
+	return &NetworkSplitter{
+		networks: ns.networks,
+	}
+}
+
+// Reset resets the internal position of the iterator to the start of the collection.
+func (ns *NetworkSplitter) Reset() {
+	ns.index = 0
+	ns.offset = 0
+}
+
+// IPverSubset returns a new NetworkSplitter containing only IPv4 or IPv6 networks depending on ipVersion parameter.
+func (ns *NetworkSplitter) IPverSubset(ipVersion IPVersion) *NetworkSplitter {
+	newNS := &NetworkSplitter{
+		networks: make([]*NetworkToSplit, 0),
+	}
+
+	for _, nw := range ns.networks {
+		if nw.ipVersion == ipVersion {
+			newNS.networks = append(newNS.networks, nw)
 		}
-		localPools = append(localPools, splitNetwork(p.Size, b)...)
 	}
-	return localPools, nil
+
+	return newNS
+}
+
+// Next indicates whether there's at least one subnet from its NetworkToSplit the iterator didn't return so far.
+// It shall be called before Get to make sure the iterator didn't reach the end.
+func (ns *NetworkSplitter) Next() bool {
```

Comment:
I think we could have `Get()` return `nil` when it is used up instead of having this extra function.
`Next` itself is a bit of a misnomer since it's not advancing anything just checking if `Get()` would return something.
It seems like all the call sites could do `for net := splitter.Get(); net != nil { ... }` instead.

Reply:
Although in most cases checking the return value of `Get` would do the job, I think it'd make the code resetting the iterator harder to understand (cf. [see here](https://github.com/moby/moby/pull/43033/files#diff-9265e3cef0c7e0f9b3373f21ede0e1a56f625a8d86785655e1229636f9337979R432-R435)). However, I think you're right: `Next` is badly named. Maybe I can rename it into `HasNext` or something like that. What do you think?

Reply:
It just feels better to let `Get` return nil if there's no more subnet to return. As such, I renamed `Next` into `HasNext` to make it more convenient to check if the splitter needs to be reset and removed all calls to `Next` except for that particular case. Let me know if you have other suggestions here.